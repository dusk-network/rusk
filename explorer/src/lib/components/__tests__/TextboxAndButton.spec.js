import { afterEach, describe, expect, it, vi } from "vitest";
import { mdiMagnify } from "@mdi/js";
import { cleanup, fireEvent, render } from "@testing-library/svelte";
import { TextboxAndButton } from "..";

describe("TextField", () => {
  afterEach(cleanup);

  /** @type {import("svelte").ComponentProps<TextboxAndButton>} */
  const baseProps = {
    className: "foo bar",
    icon: {
      path: mdiMagnify,
      position: "after",
      size: "default",
    },
    placeholder: "Search",
  };

  const baseOptions = {
    props: baseProps,
    target: document.body,
  };

  it("renders the component with default values", async () => {
    const { getByRole, container } = render(TextboxAndButton, baseOptions);

    const input = getByRole("textbox");
    const button = getByRole("button");

    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("updates input value on change", async () => {
    const { getByRole, container } = render(TextboxAndButton, baseOptions);
    const input = getByRole("textbox");

    await fireEvent.input(input, { target: { value: "test value" } });

    expect(input).toHaveValue("test value");

    expect(container.firstChild).toMatchSnapshot();
  });

  it("triggers click event on button click", async () => {
    const { getByRole, component } = render(TextboxAndButton, baseOptions);
    const button = getByRole("button");

    const mockClickHandler = vi.fn();
    component.$on("click", mockClickHandler);
    await fireEvent.click(button);

    expect(mockClickHandler).toHaveBeenCalledOnce();
  });

  it("focuses input on focus() call", async () => {
    const { getByRole, component } = render(TextboxAndButton);
    const input = getByRole("textbox");

    component.focus();
    expect(input).toHaveFocus();
  });

  it("should expose a method to select the element's text", () => {
    const { component, getByRole } = render(TextboxAndButton, {
      ...baseProps,
      value: "some input text",
    });

    const input = /** @type {HTMLInputElement} */ (getByRole("textbox"));

    component.select();

    const selectedText = input.value.substring(
      Number(input.selectionStart),
      Number(input.selectionEnd)
    );

    expect(selectedText).toBe("some input text");
  });
});
