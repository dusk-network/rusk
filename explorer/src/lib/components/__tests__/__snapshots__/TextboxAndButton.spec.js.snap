// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField > renders the component with default values 1`] = `
<div
  class="textbox-button foo bar"
  style="width: 250px;"
>
  <input
    class="dusk-textbox dusk-textbox-text textbox-button__input"
    placeholder="Search"
    required=""
    type="text"
  />
  
   
  <button
    class="dusk-button dusk-button--type--submit dusk-button--variant--primary dusk-button--size--default dusk-icon-button"
    type="submit"
  >
     
    <svg
      class="dusk-icon dusk-icon--size--default dusk-button__icon"
      role="graphics-symbol"
      viewBox="0 0 24 24"
    >
      <path
        d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z"
      />
    </svg>
    
  </button>
</div>
`;

exports[`TextField > updates input value on change 1`] = `
<div
  class="textbox-button foo bar"
  style="width: 250px;"
>
  <input
    class="dusk-textbox dusk-textbox-text textbox-button__input"
    placeholder="Search"
    required=""
    type="text"
  />
  
   
  <button
    class="dusk-button dusk-button--type--submit dusk-button--variant--primary dusk-button--size--default dusk-icon-button"
    type="submit"
  >
     
    <svg
      class="dusk-icon dusk-icon--size--default dusk-button__icon"
      role="graphics-symbol"
      viewBox="0 0 24 24"
    >
      <path
        d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z"
      />
    </svg>
    
  </button>
</div>
`;
